File: frontend/services/onboarding.ts
import { apiFetch } from "./api/client";

const BASE = "/api/v1/onboarding";

export interface OnboardingSession {
  session_id: string;
  current_step: number;
  status: string;
  company_name?: string | null;
  industry_code?: string | null;
  industry_subcategory?: string | null;
  estimated_locations?: number | null;
  brand_color?: string | null;
  logo_url?: string | null;
  website_url?: string | null;
  employee_source?: string | null;
  launch_progress?: Record<string, unknown> | null;
}

export interface CompanyProfile {
  company_name: string;
  industry_code: string;
  industry_subcategory?: string | null;
  estimated_locations?: number | null;
  brand_color_hex?: string | null;
  logo_url?: string | null;
  confidence: number;
}

export interface TemplateItem {
  id: string;
  category: string;
  name: string;
  description?: string | null;
  is_recommended: boolean;
  is_selected: boolean;
  content_preview: Record<string, unknown>;
}

export interface TemplateCategoryGroup {
  category: string;
  display_name: string;
  icon: string;
  items: TemplateItem[];
  selected_count: number;
  total_count: number;
}

export interface IndustryPackage {
  package_name: string;
  industry_code: string;
  categories: TemplateCategoryGroup[];
  total_selected: number;
  total_available: number;
}

export interface SelectionSummary {
  forms: number;
  checklists: number;
  audits: number;
  issue_categories: number;
  workflows: number;
  training_modules: number;
  shift_templates: number;
  repair_manuals: number;
  badges: number;
  total_selected: number;
  total_available: number;
}

export interface OnboardingLocation {
  id?: string;
  name: string;
  address?: string | null;
}

export interface OnboardingAsset {
  id?: string;
  name: string;
  category: string;
  model?: string | null;
  manufacturer?: string | null;
  location_name?: string | null;
}

export interface OnboardingVendor {
  id?: string;
  name: string;
  service_type?: string | null;
  contact_email?: string | null;
  contact_phone?: string | null;
}

export interface OnboardingEmployee {
  id?: string;
  full_name: string;
  email: string;
  phone?: string;
  position?: string;
  department?: string;
  retail_role: string;
  location_name?: string;
  reports_to?: string;
  status?: string;
}

export interface RoleMapping {
  id: string;
  source_title: string;
  source_department?: string | null;
  retail_role: string;
  confidence_score: number;
  is_confirmed: boolean;
  employee_count: number;
  low_confidence: boolean;
}

export interface WorkspacePreview {
  summary: SelectionSummary;
  locations: Record<string, unknown>[];
  assets: Record<string, unknown>[];
  vendors: Record<string, unknown>[];
  forms_and_checklists: Record<string, unknown>[];
  issue_categories: Record<string, unknown>[];
  workflows: Record<string, unknown>[];
  training_modules: Record<string, unknown>[];
  shift_templates: Record<string, unknown>[];
  repair_manuals: Record<string, unknown>[];
  employees: Record<string, unknown>;
  company_name?: string | null;
  brand_color?: string | null;
  logo_url?: string | null;
}

export interface LaunchProgress {
  status: string;
  current_step?: string | null;
  progress_percent: number;
  steps_completed: string[];
  steps_remaining: string[];
  error?: string | null;
}

// ── Session ──────────────────────────────────────────────────────────────────

export const createSession = (): Promise<OnboardingSession> =>
  apiFetch(`${BASE}/sessions`, { method: "POST" });

export const getCurrentSession = (): Promise<OnboardingSession | null> =>
  apiFetch<OnboardingSession>(`${BASE}/sessions/current`).catch(() => null);

// ── Step 1 ───────────────────────────────────────────────────────────────────

export const discoverCompany = (sessionId: string, websiteUrl: string): Promise<CompanyProfile> =>
  apiFetch(`${BASE}/sessions/${sessionId}/discover`, {
    method: "POST",
    body: JSON.stringify({ website_url: websiteUrl }),
  });

export const discoverFallback = (
  sessionId: string,
  payload: { company_name: string; industry_code: string; estimated_locations?: number }
): Promise<CompanyProfile> =>
  apiFetch(`${BASE}/sessions/${sessionId}/discover/fallback`, {
    method: "POST",
    body: JSON.stringify(payload),
  });

export const confirmCompany = (sessionId: string, profile: CompanyProfile): Promise<OnboardingSession> =>
  apiFetch(`${BASE}/sessions/${sessionId}/confirm-company`, {
    method: "POST",
    body: JSON.stringify(profile),
  });

// ── Step 2 ───────────────────────────────────────────────────────────────────

export const getTemplates = (sessionId: string): Promise<IndustryPackage> =>
  apiFetch(`${BASE}/sessions/${sessionId}/templates`);

export const updateSelections = (
  sessionId: string,
  updates: { template_id: string; is_selected: boolean }[]
): Promise<{ ok: boolean }> =>
  apiFetch(`${BASE}/sessions/${sessionId}/selections`, {
    method: "PATCH",
    body: JSON.stringify(updates),
  });

export const getSelectionSummary = (sessionId: string): Promise<SelectionSummary> =>
  apiFetch(`${BASE}/sessions/${sessionId}/selections/summary`);

export const confirmTemplates = (sessionId: string): Promise<OnboardingSession> =>
  apiFetch(`${BASE}/sessions/${sessionId}/confirm-templates`, { method: "POST" });

// ── Step 3: Locations ─────────────────────────────────────────────────────────

export const suggestLocations = (sessionId: string): Promise<OnboardingLocation[]> =>
  apiFetch(`${BASE}/sessions/${sessionId}/suggest-locations`);

export const addLocation = (sessionId: string, loc: OnboardingLocation): Promise<OnboardingLocation> =>
  apiFetch(`${BASE}/sessions/${sessionId}/locations`, {
    method: "POST",
    body: JSON.stringify(loc),
  });

export const listLocations = (sessionId: string): Promise<OnboardingLocation[]> =>
  apiFetch(`${BASE}/sessions/${sessionId}/locations`);

export const deleteLocation = (sessionId: string, locId: string): Promise<void> =>
  apiFetch(`${BASE}/sessions/${sessionId}/locations/${locId}`, { method: "DELETE" });

export const confirmLocations = (sessionId: string): Promise<OnboardingSession> =>
  apiFetch(`${BASE}/sessions/${sessionId}/confirm-locations`, { method: "POST" });

// ── Step 4: Assets & Vendors ──────────────────────────────────────────────────

export const suggestAssets = (sessionId: string): Promise<OnboardingAsset[]> =>
  apiFetch(`${BASE}/sessions/${sessionId}/suggest-assets`);

export const addAsset = (sessionId: string, asset: OnboardingAsset): Promise<OnboardingAsset> =>
  apiFetch(`${BASE}/sessions/${sessionId}/assets`, {
    method: "POST",
    body: JSON.stringify(asset),
  });

export const listAssets = (sessionId: string): Promise<OnboardingAsset[]> =>
  apiFetch(`${BASE}/sessions/${sessionId}/assets`);

export const deleteAsset = (sessionId: string, assetId: string): Promise<void> =>
  apiFetch(`${BASE}/sessions/${sessionId}/assets/${assetId}`, { method: "DELETE" });

export const addVendor = (sessionId: string, vendor: OnboardingVendor): Promise<OnboardingVendor> =>
  apiFetch(`${BASE}/sessions/${sessionId}/vendors`, {
    method: "POST",
    body: JSON.stringify(vendor),
  });

export const listVendors = (sessionId: string): Promise<OnboardingVendor[]> =>
  apiFetch(`${BASE}/sessions/${sessionId}/vendors`);

export const deleteVendor = (sessionId: string, vendorId: string): Promise<void> =>
  apiFetch(`${BASE}/sessions/${sessionId}/vendors/${vendorId}`, { method: "DELETE" });

export const confirmAssets = (sessionId: string): Promise<OnboardingSession> =>
  apiFetch(`${BASE}/sessions/${sessionId}/confirm-assets`, { method: "POST" });

export const confirmVendors = (sessionId: string): Promise<OnboardingSession> =>
  apiFetch(`${BASE}/sessions/${sessionId}/confirm-vendors`, { method: "POST" });

// ── Step 5: Team ──────────────────────────────────────────────────────────────

export const setEmployeeSource = (sessionId: string, source: string): Promise<OnboardingSession> =>
  apiFetch(`${BASE}/sessions/${sessionId}/employee-source`, {
    method: "POST",
    body: JSON.stringify({ source }),
  });

export const addEmployee = (sessionId: string, employee: OnboardingEmployee): Promise<{ id: string }> =>
  apiFetch(`${BASE}/sessions/${sessionId}/employees`, {
    method: "POST",
    body: JSON.stringify(employee),
  });

export const listEmployees = (sessionId: string): Promise<{ employees: OnboardingEmployee[]; total: number }> =>
  apiFetch(`${BASE}/sessions/${sessionId}/employees`);

export const deleteEmployee = (sessionId: string, employeeId: string): Promise<void> =>
  apiFetch(`${BASE}/sessions/${sessionId}/employees/${employeeId}`, { method: "DELETE" });

export const generateInviteLink = (
  sessionId: string,
  config?: { default_role?: string; expiry_hours?: number }
): Promise<{ invite_url: string; qr_code_data: string; expires_at: string }> =>
  apiFetch(`${BASE}/sessions/${sessionId}/invite-link`, {
    method: "POST",
    body: JSON.stringify(config ?? {}),
  });

export const getRoleMappings = (sessionId: string): Promise<RoleMapping[]> =>
  apiFetch(`${BASE}/sessions/${sessionId}/role-mappings`);

export const confirmEmployees = (sessionId: string): Promise<OnboardingSession> =>
  apiFetch(`${BASE}/sessions/${sessionId}/confirm-employees`, { method: "POST" });

// ── Step 6: Shift Settings ────────────────────────────────────────────────────

export const confirmShiftSettings = (sessionId: string): Promise<OnboardingSession> =>
  apiFetch(`${BASE}/sessions/${sessionId}/confirm-shift-settings`, { method: "POST" });

// ── Step 7: Preview ───────────────────────────────────────────────────────────

export const getWorkspacePreview = (sessionId: string): Promise<WorkspacePreview> =>
  apiFetch(`${BASE}/sessions/${sessionId}/preview`);

export const confirmPreview = (sessionId: string): Promise<OnboardingSession> =>
  apiFetch(`${BASE}/sessions/${sessionId}/confirm-preview`, { method: "POST" });

// ── Step 7: Launch ────────────────────────────────────────────────────────────

export const launchWorkspace = (sessionId: string): Promise<{ success: boolean; message: string }> =>
  apiFetch(`${BASE}/sessions/${sessionId}/launch`, { method: "POST" });

export const getLaunchProgress = (sessionId: string): Promise<LaunchProgress> =>
  apiFetch(`${BASE}/sessions/${sessionId}/launch-progress`);

export const getFirstActions = (
  sessionId: string
): Promise<{ actions: { title: string; description: string; icon: string; action_url: string; action_label: string }[] }> =>
  apiFetch(`${BASE}/sessions/${sessionId}/first-actions`);

// ── Package templates (post-onboarding) ─────────────────────────────────────

export interface PackageTemplateItem {
  id: string;
  name: string;
  description: string;
  category: string;
  content: Record<string, unknown>;
  sort_order: number;
}

export interface PackageTemplatesResponse {
  items: PackageTemplateItem[];
  industry_code: string | null;
}

export const getPackageTemplates = (category?: string): Promise<PackageTemplatesResponse> => {
  const qs = category ? `?category=${encodeURIComponent(category)}` : "";
  return apiFetch(`${BASE}/package-templates${qs}`);
};

File: backend/models/onboarding.py
from pydantic import BaseModel
from typing import Optional, List, Any
from enum import Enum


class IndustryCode(str, Enum):
    QSR = "qsr"
    RETAIL_FASHION = "retail_fashion"
    RETAIL_GROCERY = "retail_grocery"
    HOSPITALITY = "hospitality"
    HEALTHCARE_CLINIC = "healthcare_clinic"
    MANUFACTURING = "manufacturing"
    LOGISTICS = "logistics"
    CASUAL_DINING = "casual_dining"
    FULL_SERVICE_RESTAURANT = "full_service_restaurant"
    CAFE_BAR = "cafe_bar"
    BAKERY = "bakery"


INDUSTRY_DISPLAY = {
    "qsr": "Quick Service Restaurant",
    "retail_fashion": "Retail — Fashion & Apparel",
    "retail_grocery": "Retail — Grocery & Convenience",
    "hospitality": "Hospitality & Hotels",
    "healthcare_clinic": "Healthcare — Clinics",
    "manufacturing": "Manufacturing",
    "logistics": "Logistics & Warehousing",
    "casual_dining": "Casual Dining Restaurant",
    "full_service_restaurant": "Full-Service Restaurant",
    "cafe_bar": "Cafe & Bar",
    "bakery": "Bakery & Pastry",
}


# ── Step 1: Company Discovery ──────────────────────────────────────────────────

class CompanyProfile(BaseModel):
    company_name: str
    industry_code: str
    industry_subcategory: Optional[str] = None
    estimated_locations: Optional[int] = None
    brand_color_hex: Optional[str] = None
    logo_url: Optional[str] = None
    confidence: float = 0.0


class CompanyDiscoveryRequest(BaseModel):
    website_url: str


class CompanyDiscoveryFallbackRequest(BaseModel):
    company_name: str
    industry_code: str
    industry_subcategory: Optional[str] = None
    estimated_locations: Optional[int] = None


class OnboardingSessionResponse(BaseModel):
    session_id: str
    current_step: int
    status: str
    company_name: Optional[str] = None
    industry_code: Optional[str] = None
    industry_subcategory: Optional[str] = None
    estimated_locations: Optional[int] = None
    brand_color: Optional[str] = None
    logo_url: Optional[str] = None
    website_url: Optional[str] = None
    employee_source: Optional[str] = None
    launch_progress: Optional[dict] = None


# ── Step 2: Template Selection ─────────────────────────────────────────────────

class TemplateItemResponse(BaseModel):
    id: str
    category: str
    name: str
    description: Optional[str] = None
    is_recommended: bool
    is_selected: bool
    content_preview: dict


class TemplateCategoryGroup(BaseModel):
    category: str
    display_name: str
    icon: str
    items: List[TemplateItemResponse]
    selected_count: int
    total_count: int


class IndustryPackageResponse(BaseModel):
    package_name: str
    industry_code: str
    categories: List[TemplateCategoryGroup]
    total_selected: int
    total_available: int


class SelectionUpdate(BaseModel):
    template_id: str
    is_selected: bool


class SelectionSummary(BaseModel):
    forms: int = 0
    checklists: int = 0
    audits: int = 0
    issue_categories: int = 0
    workflows: int = 0
    training_modules: int = 0
    shift_templates: int = 0
    repair_manuals: int = 0
    badges: int = 0
    total_selected: int = 0
    total_available: int = 0


# ── Step 3: Locations ─────────────────────────────────────────────────────────

class OnboardingLocation(BaseModel):
    id: Optional[str] = None
    name: str
    address: Optional[str] = None


# ── Step 4: Assets & Vendors ──────────────────────────────────────────────────

class OnboardingAsset(BaseModel):
    id: Optional[str] = None
    name: str
    category: str
    model: Optional[str] = None
    manufacturer: Optional[str] = None
    location_name: Optional[str] = None


class OnboardingVendor(BaseModel):
    id: Optional[str] = None
    name: str
    service_type: Optional[str] = None
    contact_email: Optional[str] = None
    contact_phone: Optional[str] = None


# ── Step 5: Employee Setup ─────────────────────────────────────────────────────

class EmployeeSourceRequest(BaseModel):
    source: str  # sprout_hr | hris_other | csv | manual | invite_link


class ManualEmployeeInput(BaseModel):
    full_name: str
    email: str
    phone: Optional[str] = None
    position: Optional[str] = None
    department: Optional[str] = None
    retail_role: str = "staff"
    location_name: Optional[str] = None


class InviteConfig(BaseModel):
    default_role: str = "staff"
    location_id: Optional[str] = None
    expiry_hours: int = 72


class InviteResult(BaseModel):
    invite_url: str
    qr_code_data: str  # base64 PNG
    expires_at: str


class RoleMappingResponse(BaseModel):
    id: str
    source_title: str
    source_department: Optional[str] = None
    retail_role: str
    confidence_score: float
    is_confirmed: bool
    employee_count: int
    low_confidence: bool  # True if < 0.7


class RoleMappingUpdate(BaseModel):
    retail_role: str


class CSVImportResult(BaseModel):
    total_rows: int
    valid_rows: int
    error_rows: int
    errors: List[dict] = []
    import_job_id: str


class EmployeeImportSummary(BaseModel):
    total_employees: int
    by_role: dict
    pending_role_assignment: int
    import_source: Optional[str] = None


# ── Step 6: Workspace Preview ─────────────────────────────────────────────────

class WorkspacePreview(BaseModel):
    summary: SelectionSummary
    locations: List[dict] = []
    assets: List[dict] = []
    vendors: List[dict] = []
    forms_and_checklists: List[dict] = []
    issue_categories: List[dict] = []
    workflows: List[dict] = []
    training_modules: List[dict] = []
    shift_templates: List[dict] = []
    repair_manuals: List[dict] = []
    employees: dict = {}
    company_name: Optional[str] = None
    brand_color: Optional[str] = None
    logo_url: Optional[str] = None


# ── Step 7: Launch ─────────────────────────────────────────────────────────────

class LaunchProgress(BaseModel):
    status: str  # pending | provisioning | completed | failed
    current_step: Optional[str] = None
    progress_percent: int = 0
    steps_completed: List[str] = []
    steps_remaining: List[str] = []
    error: Optional[str] = None


class GuidedAction(BaseModel):
    title: str
    description: str
    icon: str
    action_url: str
    action_label: str


class LaunchResult(BaseModel):
    success: bool
    message: str
    entities_created: dict = {}
